stringfuzz/transformers/graft.py

```diff
@@ -0,0 +1,129 @@
+'''
+The graft transform picks a subtree and a leaf at random
+and swaps them for each type.
+'''
+
+import random
+
+from stringfuzz.ast import *
+from stringfuzz.types import *
+from stringfuzz.ast_walker import ASTWalker
+from stringfuzz.generators import random_text
+from stringfuzz.parser import parse
+
+__all__ = [
+    'graft',
+]
+
+class GraftTransformer(ASTWalker):
+    def __init__(self, ast, pairs):
+        super(GraftTransformer, self).__init__(ast)
+        self.pairs = pairs
+
+    @property
+    def ast(self):
+        return self._ASTWalker__ast
+
+    def enter_expression(self, expr):
+        for i in range(len(expr.body)):
+            for pair in self.pairs:
+                if expr.body[i] == pair[0]:
+                    expr.body[i] = pair[1]
+                elif expr.body[i] == pair[1]:
+                    expr.body[i] = pair[0]
+
+class GraftFinder(ASTWalker):
+    def __init__(self, ast):
+        super(GraftFinder, self).__init__(ast)
+        #            expr, lit
+        self.str  = [None, None]
+        self.bool = [None, None]
+        self.int  = [None, None]
+        self.rx   = [None, None]
+
+    @property
+    def pairs(self):
+        pairs = []
+        if all(self.str):
+            pairs.append(self.str)
+        if all(self.bool):
+            pairs.append(self.bool)
+        if all(self.int):
+            pairs.append(self.int)
+        if all(self.rx):
+            pairs.append(self.rx)
+        return pairs
+
+    def enter_literal(self, literal):
+        replace = random.choice([True, False])
+        if isinstance(literal, StringLitNode):
+            if self.str[1]:
+                if replace:
+                    self.str[1] = literal
+            else:
+                self.str[1] = literal
+        elif isinstance(literal, BoolLitNode):
+            if self.bool[1]:
+                if replace:
+                    self.bool[1] = literal
+            else:
+                self.bool[1] = literal
+        elif isinstance(literal, IntLitNode):
+            if self.int[1]:
+                if replace:
+                    self.int[1] = literal
+            else:
+                self.int[1] = literal
+
+    def enter_identifier(self, ident):
+        #TODO How to check type of identifiers?
```

Comment:
The only way I can think of, is to pre-process the problem and look at the declarations. What do you think of that?